src/analyzer/analyzerthread.cpp

```diff
@@ -0,0 +1,342 @@
+#include "analyzer/analyzerwaveform.h"
+
+#include "library/dao/analysisdao.h"
+
+#include "engine/engine.h"
+
+#include "sources/soundsourceproxy.h"
+#include "sources/audiosourcestereoproxy.h"
+
+#include "util/db/dbconnectionpooler.h"
+#include "util/db/dbconnectionpooled.h"
+#include "util/logger.h"
+#include "util/timer.h"
+
+
+namespace {
+
+mixxx::Logger kLogger("AnalyzerThread");
+
+// Analysis is done in blocks.
+// We need to use a smaller block size, because on Linux the AnalyzerThread
+// can starve the CPU of its resources, resulting in xruns. A block size
+// of 4096 frames per block seems to do fine.
+constexpr mixxx::AudioSignal::ChannelCount kAnalysisChannels = mixxx::kEngineChannelCount;
+constexpr SINT kAnalysisFramesPerBlock = 4096;
+const SINT kAnalysisSamplesPerBlock =
+        kAnalysisFramesPerBlock * kAnalysisChannels;
+
+// Maximum frequency of progress updates while busy
+constexpr std::chrono::milliseconds kBusyProgressInhibitDuration(100);
+
+std::atomic<int> s_instanceCounter(0);
+
+} // anonymous namespace
+
+AnalyzerThread::AnalyzerThread(
+        int id,
+        mixxx::DbConnectionPoolPtr pDbConnectionPool,
+        UserSettingsPointer pConfig)
+        : m_id(id),
+          m_pDbConnectionPool(std::move(pDbConnectionPool)),
+          m_pConfig(std::move(pConfig)),
+          m_run(true),
+          m_sampleBuffer(kAnalysisSamplesPerBlock),
+          m_emittedState(AnalyzerThreadState::Void),
+          // The first signal should always be emitted
+          m_lastBusyProgressEmittedAt(Clock::now() - kBusyProgressInhibitDuration) {
+    m_nextTrack.setValue(TrackPointer());
+    m_analyzerProgress.setValue(kAnalyzerProgressUnknown);
+    // This type is registered multiple times although once would be sufficient
+    qRegisterMetaType<AnalyzerThreadState>();
+}
+
+AnalyzerThread::~AnalyzerThread() {
+    kLogger.debug() << "Destroying" << m_id;
+    VERIFY_OR_DEBUG_ASSERT(isFinished()) {
+        stop();
+        // The following operation will block the host thread!
+        wait();
+    }
+}
+
+void AnalyzerThread::run() {
+    if (!m_run.load()) {
+        return;
+    }
+
+    const int instanceId = s_instanceCounter.fetch_add(1) + 1;
+    QThread::currentThread()->setObjectName(QString("AnalyzerThread %1").arg(instanceId));
+
+    kLogger.debug() << "Running" << m_id;
+
+    exec();
+
+    kLogger.debug() << "Exiting" << m_id;
+
+    m_run.store(false);
+
+    emitProgress(AnalyzerThreadState::Void);
+}
+
+void AnalyzerThread::exec() {
+    // The thread-local database connection for waveform analysis
+    // must not be closed before returning from this function.
+    // Therefore the DbConnectionPooler is defined here independent
+    // of whether a database connection will be opened or not.
+    mixxx::DbConnectionPooler dbConnectionPooler;
+
+    std::unique_ptr<AnalysisDao> pAnalysisDao;
+    if (m_pConfig->getValue<bool>(ConfigKey("[Library]", "EnableWaveformGenerationWithAnalysis"), true)) {
```

Comment:
This was a simple refactoring that didn't change the original meaning of the code. If the code is and was wrong than it should be fixed.

Reply:
You did change the meaning, because you removed the AnalyzerMode parameter.
Previously, AnalysisFeature indicated that the option could be followed, and PlayerManager that it should be ignored.

Reply:
The refactoring really changed the behaviour! Moving this code away from AnalysisFeature as @daschuer suggested was a bad idea ;)

Reply:
FYI, the purpose of the option is to save disk space, not time for batch analysis.